compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2693,46 +2699,126 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
-                            .map(|p| {
-                                p.join(if sess.host.is_like_windows { "ld64.exe" } else { "ld64" })
-                            })
-                            .find(|p| p.exists())
-                            .unwrap_or_else(|| sess.fatal("rust-lld (as ld64) not found"));
-                        cmd.cmd().arg({
-                            let mut arg = OsString::from("-fuse-ld=");
-                            arg.push(ld64_exe);
-                            arg
-                        });
-                    } else {
-                        let tools_path = sess.get_tools_search_paths(false);
-                        let lld_path = tools_path
-                            .into_iter()
-                            .map(|p| p.join("gcc-ld"))
-                            .find(|p| {
-                                p.join(if sess.host.is_like_windows { "ld.exe" } else { "ld" })
-                                    .exists()
-                            })
-                            .unwrap_or_else(|| sess.fatal("rust-lld (as ld) not found"));
-                        cmd.cmd().arg({
-                            let mut arg = OsString::from("-B");
-                            arg.push(lld_path);
-                            arg
-                        });
-                    }
-                }
+/// This takes care of the various possible enrichments to the linking that can be requested on the
+/// CLI (and emitting errors and warnings when applicable):
+/// - shortcuts to `-fuse-ld` with the `gcc` flavor
+/// - the unstable `-Zgcc-ld=lld` flag to use `rust-lld`, stabilized as the following item
+fn handle_cli_linker_flavors(
+    cmd: &mut dyn Linker,
+    sess: &Session,
+    flavor: LinkerFlavor,
+    crt_objects_fallback: bool,
+) {
+    let unstable_gcc_lld = sess.opts.debugging_opts.gcc_ld == Some(LdImpl::Lld);
+    if unstable_gcc_lld {
+        // Sanity check: ensure `gcc` is the currently selected flavor.
+        if LinkerFlavor::Gcc != flavor {
+            sess.fatal("`-Zgcc-ld` is used even though the linker flavor is not `gcc`");
+        }
+    }
+
+    let cg = &sess.opts.cg;
+
+    // The `-C linker-flavor` CLI flag can optionally enrich linker-flavors. Check whether that's
+    // applicable, and emit errors if sanity checks fail. There's currently only one enrichment:
+    // adding an argument to the `cc` invocation to use the `use_ld` given linker.
+    let use_ld = match &cg.linker_flavor {
+        Some(LinkerFlavorCli::Gcc { use_ld }) => {
+            // Ensure `gcc` is the currently selected flavor. Error out cleanly, as `-Zgcc-ld` does
+            // if that happens, but this should be unreachable.
+            if LinkerFlavor::Gcc != flavor {
+                sess.fatal(
+                    "`-Clinker-flavor=gcc:*` flag is used even though the \
+                    linker flavor is not `gcc`",
+                );
```

Comment:
How can this happen? They should always be in sync by definition, right?

Reply:
It's not supposed to indeed, as long as this piece of code stays downstream from the CLI linker-flavor lowering, and the actual `LinkerFlavor` is not changed or overriden by the time the order-independent options are added to the cmd args. If this happens by mistake (or intentionally) in the future, this check will catch it, but it's not going to trigger today.

Reply:
Maybe change it to a `bug!()` instead?

Reply:
Good idea Thank you.

I've pushed a commit to address this.